src/contracts/AbstractARM.sol

```diff
@@ -0,0 +1,686 @@
+
+        amounts = new uint256[](2);
+        amounts[0] = amountIn;
+        amounts[1] = amountOut;
+    }
+
+    /**
+     * @notice Receive an exact amount of output tokens for as few input tokens as possible.
+     * msg.sender should have already given the router an allowance of
+     * at least amountInMax on the input token.
+     *
+     * @param inToken Input token.
+     * @param outToken Output token.
+     * @param amountOut The amount of output tokens to receive.
+     * @param amountInMax The maximum amount of input tokens that can be required before the transaction reverts.
+     * @param to Recipient of the output tokens.
+     */
+    function swapTokensForExactTokens(
+        IERC20 inToken,
+        IERC20 outToken,
+        uint256 amountOut,
+        uint256 amountInMax,
+        address to
+    ) external virtual {
+        uint256 amountIn = _swapTokensForExactTokens(inToken, outToken, amountOut, to);
+
+        require(amountIn <= amountInMax, "ARM: Excess input amount");
+    }
+
+    /**
+     * @notice Uniswap V2 Router compatible interface. Receive an exact amount of
+     * output tokens for as few input tokens as possible.
+     * msg.sender should have already given the router an allowance of
+     * at least amountInMax on the input token.
+     *
+     * @param amountOut The amount of output tokens to receive.
+     * @param amountInMax The maximum amount of input tokens that can be required before the transaction reverts.
+     * @param path The input and output token addresses.
+     * @param to Recipient of the output tokens.
+     * @param deadline Unix timestamp after which the transaction will revert.
+     * @return amounts The input and output token amounts.
+     */
+    function swapTokensForExactTokens(
+        uint256 amountOut,
+        uint256 amountInMax,
+        address[] calldata path,
+        address to,
+        uint256 deadline
+    ) external virtual returns (uint256[] memory amounts) {
+        require(path.length == 2, "ARM: Invalid path length");
+        _inDeadline(deadline);
+
+        IERC20 inToken = IERC20(path[0]);
+        IERC20 outToken = IERC20(path[1]);
+
+        uint256 amountIn = _swapTokensForExactTokens(inToken, outToken, amountOut, to);
+
+        require(amountIn <= amountInMax, "ARM: Excess input amount");
+
+        amounts = new uint256[](2);
+        amounts[0] = amountIn;
+        amounts[1] = amountOut;
+    }
+
+    function _inDeadline(uint256 deadline) internal view {
+        require(deadline >= block.timestamp, "ARM: Deadline expired");
+    }
+
+    /// @dev Ensure any liquidity assets reserved for the withdrawal queue are not used
+    /// in swaps that send liquidity assets out of the ARM
+    function _transferAsset(address asset, address to, uint256 amount) internal virtual {
+        if (asset == liquidityAsset) {
+            require(amount <= _liquidityAvailable(), "ARM: Insufficient liquidity");
```

Comment:
It's probably more gas efficient for swaps of WETH out of the ARM, which is most of them, if the liquidity (WETH) available was in storage and not dynamically calculated as it currently is in  `_liquidityAvailable`. Especially since `_liquidityAvailable` does an external call to WETH to get the balance of WETH in the ARM.

It means we need to maintain another balance in storage but that's outside the swap functions.

I'll experiment and see what the savings are.

Reply:
Actually, swaps of WETH into the ARM would have to write to the new `liquidityAvailable` storage variable but that's not the typical swap of the ARM. Most of the time the ARM is buying stETH which means it is transferring WETH out.